fix recursion error when getting an invalid action object attribute

ryankwilliams opened the pull request:
This commit updates the ActionContainer __getattr__ method to not use
hasattr to determine if the attribute does not exist. When running
with Python 3, this was causing the method to call itself since
hasattr calls getattr. This update will remove using hasattr and
now use the in keyword. Now if action getattr is called with an
invalid attribute, an attribute error will be rasied.

Closes #33

ryankwilliams committed:
fix recursion error when getting an invalid action object attribute

This commit updates the ActionContainer __getattr__ method to not use
hasattr to determine if the attribute does not exist. When running
with Python 3, this was causing the method to call itself since
hasattr calls getattr. This update will remove using hasattr and
now use the in keyword. Now if action getattr is called with an
invalid attribute, an attribute error will be rasied.

Closes #33

simon3z commented:
cc @cben

cben reviewed (commented):
Nice catch, good fix :+1:
I'm surprised it worked at all, I thought Python 2 hasattr also calls getattr.

Is it easy to add a test?

cben commented on src/manageiq_client/api.py:
```diff
@@ -587,7 +587,7 @@ def collection(self):
-        if not hasattr(self, attr):
+        if attr not in self.__dict__:
             raise AttributeError("No such action {}".format(attr))
         return getattr(self, attr)
```
consider rewriting this line too as `self.__dict__[attr]` to explicitly avoid recursion.
(not really necessary, `getattr` is not supposed to call `__getattr__` if the attribute exists, just for readabily)

ryankwilliams replied:
+1 I agree. I will make the requested change and update the PR.

ryankwilliams committed:
remove using built-in getattr to return value when getattr called

This will avoid any recursion from happening. It also provides better
readability.

ryankwilliams committed:
new test to verify error is raised when getting invalid action attribute

This test will attempt to get an invalid attribute from a
ActionContainer object and checks the exception message.

cben commented:
@simon3z please review/merge

simon3z commented:
@cben I think someone from core should pick these up (cc @chessbyte @gtanzillo @Fryguy)

cben commented:
@abellotti Now that you have merge powers, please review, this is sitting unmerged...

mkoura commented:
:+1: LGTM, merging since it's already approved by @cben
